server/util/grpc_client/grpc_client.go

```diff
@@ -216,3 +223,64 @@ func CommonGRPCClientOptions() []grpc.DialOption {
 		}),
 	}
 }
+
+type grpcClientConnPoolCache struct {
+	env         environment.Env
+	connMutex   *sync.RWMutex
+	connPoolMap map[string]interfaces.ClosableClientConn
+}
+
+func RegisterConnPoolCache(env environment.Env) {
+	c := NewGrpcClientConnPoolCache(env)
+	env.SetGrpcClientConnPoolCache(c)
+}
+
+func (cpc *grpcClientConnPoolCache) GetGrpcClientConnPoolForURL(target string) (conn interfaces.ClosableClientConn, shouldClose bool, err error) {
+	cpc.connMutex.RLock()
+	connPool, ok := cpc.connPoolMap[target]
```

Comment:
i think you are overcomplicating the locking here. Dialing, by default, is a non-blocking operation. You should be okay to take a lock here for the duration of the function.

Reply:
To be clear, I just did this for the sake of avoiding making two pools for the same connection.  There's a secondary thing where I was worried about having different threads all wait in line to grab a lock just to get a connection, but Tyler showed me some benchmarks that also convinced me this probably isn't worth it.  You're just saying I should use a plain mutex for the whole call, right?

Reply:
> You're just saying I should use a plain mutex for the whole call, right?

Yeah, you'll be fine just grabbing the lock for the duration of the entire function. It will make the code a lot simpler.

Reply:
Done